diff-containers/test/Test/Data/Map/Diff/Strict.hs

```diff
@@ -1,66 +1,100 @@
-      testGroupWithProxy (Proxy @(DiffEntry (Smaller Int))) [
-      ]
-    , testGroupWithProxy (Proxy @(DiffHistory (Smaller Int))) [
+      localOption (QuickCheckTests 1000) $
+      testGroupWithProxy (Proxy @(DiffHistory (OftenSmall Int))) [
           testSemigroupLaws
         , testMonoidLaws
         , testGroupLaws
         ]
-    , testGroupWithProxy (Proxy @(Auto (DiffHistory (Smaller Int)))) [
-          testSemigroupoidLaws
-        , testGroupoidLaws
-        ]
-    , testGroupWithProxy (Proxy @(Diff (Smaller Int) (Smaller Int))) [
+    , localOption (QuickCheckTests 1000) $
+      testGroupWithProxy (Proxy @(Diff (OftenSmall Int) (OftenSmall Int))) [
           testSemigroupLaws
         , testMonoidLaws
         , testGroupLaws
         ]
-    , testGroupWithProxy (Proxy @(Auto (Diff (Smaller Int) (Smaller Int)))) [
-          testSemigroupoidLaws
-        , testGroupoidLaws
-        ]
-    , testProperty "prop_diffThenApply @(Smaller Int)" $
-        prop_diffThenApply @(Smaller Int) @(Smaller Int)
-    , testProperty "prop_diffThenApply @Int" $
-        prop_diffThenApply @Int @Int
+    , localOption (QuickCheckTests 10000) $
+      testProperty "prop_diffingIsPositive" $
+        prop_diffingIsPositive @(OftenSmall Int) @(OftenSmall Int)
+    , localOption (QuickCheckTests 10000) $
+      testProperty "prop_diffThenApply" $
+        prop_diffThenApply @(OftenSmall Int) @(OftenSmall Int)
+    , localOption (QuickCheckTests 10000) $
+      testProperty "prop_applyMempty" $
+        prop_applyMempty @(OftenSmall Int) @(OftenSmall Int)
+    , localOption (QuickCheckMaxRatio 100) $
+      localOption (QuickCheckTests 1000) $
+      testProperty "prop_applyAllAndApplySum" $
+        prop_applyAllAndApplySum @(OftenSmall Int) @(OftenSmall Int)
+    , localOption (QuickCheckMaxRatio 100) $
+      localOption (QuickCheckTests 1000) $
+      testProperty "prop_unsafeApplyAllAndUnsafeApplySum" $
+        prop_unsafeApplyAllAndUnsafeApplySum @(OftenSmall Int) @(OftenSmall Int)
     ]
 
 {------------------------------------------------------------------------------
   Simple properties
 ------------------------------------------------------------------------------}
 
+prop_diffingIsPositive ::
+     (Ord k, Eq v)
+  => Map k v
+  -> Map k v
+  -> Property
+prop_diffingIsPositive m1 m2 = property $ isPositive (diff m1 m2)
+
 prop_diffThenApply ::
      (Show k, Show v, Ord k, Eq v)
   => Map k v
   -> Map k v
   -> Property
-prop_diffThenApply m1 m2 = applyDiff m1 (diff m1 m2) === m2
+prop_diffThenApply m1 m2 = applyDiff m1 (diff m1 m2) === Right m2
+
+prop_applyMempty ::
+     (Show k, Show v, Ord k, Eq v)
+  => Map k v
+  -> Property
+prop_applyMempty m = applyDiff m mempty === Right m
+
+prop_applyAllAndApplySum ::
+     (Show k, Show v, Ord k, Eq v)
+  => Map k v
+  -> [Diff k v]
+  -> Property
+prop_applyAllAndApplySum m ds =
+  all isPositive ds ==> foldlM applyDiff m ds === applyDiff m (mconcat ds)
```

Comment:
Do we know what percentage of tests we're discarding with the `all`? We should check this prop is not trivially satisfied.

Reply:
```shell
prop_applyAllAndApplySum:             OK (0.53s)
  +++ OK, passed 1000 tests; 50912 discarded.
prop_unsafeApplyAllAndUnsafeApplySum: OK (0.51s)
  +++ OK, passed 1000 tests; 49487 discarded.
```

A lot of tests are discarded, since the probability that the `Arbitrary` instances generate a positive diff is quite low. I've added `localOption (QuickCheckMaxRatio 100)` to the tests to ensure that we pass enough tests even though a lot of inputs are discarded